pkg/agent/plugin/svidstore/disk/disk.go

```diff
@@ -0,0 +1,311 @@
+)
+
+type certFile struct {
+	filePath string
+	pemBytes []byte
+}
+
+type diskStore struct {
+	certChain certFile
+	key       certFile
+	bundle    certFile
+}
+
+func BuiltIn() catalog.BuiltIn {
+	return builtin(New())
+}
+
+func builtin(p *DiskPlugin) catalog.BuiltIn {
+	return catalog.MakeBuiltIn(pluginName,
+		svidstorev1.SVIDStorePluginServer(p),
+		configv1.ConfigServiceServer(p),
+	)
+}
+
+func New() *DiskPlugin {
+	return &DiskPlugin{}
+}
+
+type configuration struct {
+	Directory string `hcl:"directory" json:"directory"`
+}
+
+type DiskPlugin struct {
+	svidstorev1.UnsafeSVIDStoreServer
+	configv1.UnsafeConfigServer
+
+	log         hclog.Logger
+	config      *configuration
+	trustDomain string
+	mtx         sync.RWMutex
+}
+
+// SetLogger sets the logger used by the plugin
+func (p *DiskPlugin) SetLogger(log hclog.Logger) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	p.log = log
+}
+
+// Configure configures the plugin
+func (p *DiskPlugin) Configure(ctx context.Context, req *configv1.ConfigureRequest) (*configv1.ConfigureResponse, error) {
+	config := &configuration{}
+	if err := hcl.Decode(config, req.HclConfiguration); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to decode configuration: %v", err)
+	}
+
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	p.trustDomain = req.CoreConfiguration.TrustDomain
+	p.config = config
+	if config.Directory == "" {
+		return nil, status.Error(codes.InvalidArgument, "a directory must be configured")
+	}
+
+	return &configv1.ConfigureResponse{}, nil
+}
+
+// PutX509SVID stores the specified X509-SVID in the configured location
+func (p *DiskPlugin) PutX509SVID(ctx context.Context, req *svidstorev1.PutX509SVIDRequest) (*svidstorev1.PutX509SVIDResponse, error) {
+	log := p.log.With(telemetry.SPIFFEID, req.Svid.SpiffeID)
+
+	config, err := p.getConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	diskStore, err := newDiskStore(req.Metadata, config.Directory)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	diskStore.certChain.pemBytes = certChainPEMBytes(req.Svid.CertChain)
+	diskStore.key.pemBytes = keyPEMBytes(req.Svid.PrivateKey)
+	diskStore.bundle.pemBytes = certChainPEMBytes(req.Svid.Bundle)
+
+	log.With("cert_chain_file_path", diskStore.certChain.filePath).Debug("Writing certificate chain file")
+	if err := diskStore.certChain.write(p.trustDomain); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to write certificate chain file: %v", err)
+	}
+
+	log.With("key_file_path", diskStore.key.filePath).Debug("Writing key file")
+	if err := diskStore.key.write(p.trustDomain); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to write key file: %v", err)
+	}
+
+	log.With("bundle_file_path", diskStore.bundle.filePath).Debug("Writing bundle file")
+	if err := diskStore.bundle.write(p.trustDomain); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to write bundle file: %v", err)
+	}
```

Comment:
would it make sense to best-effort delete the previously-saved components of the stored SVID if a later write action fails? This would help keep disk usage tidied up

Reply:
if we are updating a previously stored-SVID, is it possible for us to reach an inconsistent state where the past SVID is partially on disk, and the rest has been overwritten by the updated SVID?

Reply:
Thank you @amoore877 for the feedback. I'm implementing a solution that should address these concerns, providing a reasonable level of atomicity in the write action. I'll be updating the PR with that.